### PR TITLE
Sort OpenCL devices

### PR DESCRIPTION
**Pull request by claudioandre-br**
More testing is needed.

### CODE REVIEW

File: src/opencl_common.c [hunk 1]
```diff
@@ -663,10 +687,30 @@ static void add_device_to_list(int sequential_id)
...
+/* Used below (inside add_device_type routine) to sort devices */
+static int comparator(const void *p1, const void *p2)
+{
+	const speed_sort_t *c1 = (const speed_sort_t *)p1;
+	const speed_sort_t *c2 = (const speed_sort_t *)p2;
+	int diff = (int)c2->value - (int)c1->value;
+	if (diff)
+		return diff;
+	return c1->index - c2->index;
+}
```
Comment (magnumripper) [inline src/opencl_common.c]: This will ensure that devices with same speed will end up sorted by device number, lowest first, right? I like it.

Comment (claudioandre-br) [reply src/opencl_common.c]: Yep. Well, this is how JtR proper works, I just keep it.

**Comment by magnumripper** [comment]
Tried it out on my Macbook, worked fine except we still have this:
```
$ ../run/john all_tests.in -form:ntlmv2-opencl -dev=best -w
Device 2: AMD Radeon Pro 560 Compute Engine
Using default input encoding: UTF-8
Loaded 6 password hashes with 6 different salts (ntlmv2-opencl, NTLMv2 C/R [MD4 HMAC-MD5 OpenCL])
Remaining 3 password hashes with 3 different salts
Error: To fully use the 2 devices requested, you must specify --fork=2
(see doc/README-OPENCL)
```
Do we now have some variable to detect that we used "best" and not a specific list? Then we can change this to "*Warning*: To fully use the 2 devices *available*" and not bail out in that case, while keeping the message as-is and bail if I had actually said `-dev=2,1`.

Also,
```
$ ../run/john -form:ntlmv2-opencl -test=0 -dev=xx
Error: --device must be numerical, or one of "all", "cpu", "gpu" and
"acc[elerator]".
```
We need to add "best" to the list as well. I'm not sure about "top", with the above fix we may be good without even having that option coded.

**Comment by claudioandre-br** [comment]
> We need to add "best" to the list as well. I'm not sure about "top", with the above fix we may be good without even having that option coded.

I haven't added it because it seems like an hidden/internal option. To use "best", you just have to specify nothing. Anyway, if you want it, I will add.

**Comment by claudioandre-br** [comment]
> Tried it out on my Macbook, worked fine except we still have this:

How many GPUs do you have? Do you have 2 OpenCL enabled GPUs?

**Comment by magnumripper** [comment]
> > We need to add "best" to the list as well. I'm not sure about "top", with the above fix we may be good without even having that option coded.
> 
> I haven't added it because it seems like an hidden/internal option.
> 
> * To use "best", you just have to specify nothing (you have to be using using `--fork`).
> * To use "top", you just have to specify nothing (only WITHOUT `--fork`).
> 
> Anyway, if you want it, I will add.

Perhaps not, then - no need to list things not needed to type!

**Comment by magnumripper** [comment]
> > Tried it out on my Macbook, worked fine except we still have this:
> 
> How many GPUs do you have? Do you have 2 OpenCL enabled GPUs?

Yes, 1 CPU and 2 GPUs although one of the GPUs is just the Intel HD Graphics. OTOH it's not a lot slower than the AMD (this is a laptop so no monster stuff in it).

**Comment by magnumripper** [comment]
> > > > Tried it out on my Macbook, worked fine except we still have this:
> > > 
> > > 
> > > How many GPUs do you have? Do you have 2 OpenCL enabled GPUs?
> 
> > Yes, 1 CPU and 2 GPUs although one of the GPUs is just the Intel HD Graphics. OTOH it's not a lot slower than the AMD (this is a laptop so no monster stuff in it).
> 
> So, the software did the right thing. What should I do? A solution could be:
> 
> * remove "top" and
> * "best" with fork should do what it does now
> * "best" without fork should do what "top" does

Yes, almost - but on super, "best" is 5 devices. If I use it without fork, it should pick the Titan X and only give a brief warning there are more devices. But if I use it with `--fork=2` it should pick the first two from the list, `--fork=3` should pick the three first and so on.

**Comment by claudioandre-br** [comment]
Top 3 are new behavior
```
rm -f ../run/*.pot; ../run/john ~/alltests.in -form:ntlmv2-opencl -dev=best -w                  #Dev=6
rm -f ../run/*.pot; ../run/john ~/alltests.in -form:ntlmv2-opencl -fork=2 -w                       #Dev=6,7
rm -f ../run/*.pot; ../run/john ~/alltests.in -form:ntlmv2-opencl -fork=2 -w  --dev=gpu     #Dev=6,7 WARNING

rm -f ../run/*.pot; ../run/john -form:SHA256crypt-opencl alltests.in -incremental -max-run=30 -fork=2            #OK Dev=6,7
rm -f ../run/*.pot; ../run/john -form:SHA256crypt-opencl alltests.in -incremental -max-run=30 -fork=2 -dev=6     #OK Dev=6
rm -f ../run/*.pot; ../run/john -form:SHA256crypt-opencl alltests.in -incremental -max-run=30 -fork=2 -dev=6,7   #OK Dev=6,7

rm -f ../run/*.pot; ../run/john -form:SHA256crypt-opencl alltests.in -incremental -max-run=30 --fork=3 -dev=1,2  #OK Dev=1,2,1

rm -f ../run/*.pot; ../run/john -form:SHA256crypt-opencl alltests.in -incremental -max-run=30          -dev=1,2,3 #error 
rm -f ../run/*.pot; ../run/john -form:SHA256crypt-opencl alltests.in -incremental -max-run=30 -fork=2 -dev=0,1,2  #error 
rm -f ../run/*.pot; ../run/john -form:SHA256crypt-opencl alltests.in -incremental -max-run=30 --dev=gpu           #error 
```

**Pull request merged**